Fix performance issues leading to #412

quasicomputational opened the pull request:
Before, on the benchmark from #412:

```
	   dhall-hash +RTS -p -RTS

	total time  =       47.01 secs   (47007 ticks @ 1000 us, 1 processor)
	total alloc = 28,577,665,440 bytes  (excludes profiling overheads)

COST CENTRE     MODULE                     SRC                                                 %time %alloc

toList          Data.HashMap.Strict.InsOrd src/Data/HashMap/Strict/InsOrd.hs:(580,1)-(584,22)   76.3   84.5
traverseWithKey Data.HashMap.Strict.InsOrd src/Data/HashMap/Strict/InsOrd.hs:(420,1)-(421,78)    8.6    1.3
normalizeWith   Dhall.Core                 src/Dhall/Core.hs:(1338,1)-(1688,22)                  4.8    5.2
denote          Dhall.Core                 src/Dhall/Core.hs:(1260,1)-(1320,39)                  2.7    2.7
normalize       Dhall.Core                 src/Dhall/Core.hs:1234:1-41                           2.7    4.1
typeWithA       Dhall.TypeCheck            src/Dhall/TypeCheck.hs:(97,1)-(724,49)                2.1    0.0
shift           Dhall.Core                 src/Dhall/Core.hs:(660,1)-(813,29)                    1.2    1.5
```

After:

```
	   dhall-hash +RTS -p -RTS

	total time  =        4.96 secs   (4962 ticks @ 1000 us, 1 processor)
	total alloc = 3,026,587,104 bytes  (excludes profiling overheads)

COST CENTRE     MODULE                     SRC                                                 %time %alloc

typeWithA       Dhall.TypeCheck            src/Dhall/TypeCheck.hs:(98,1)-(737,49)               25.9   11.6
normalizeWith   Dhall.Core                 src/Dhall/Core.hs:(1326,1)-(1676,22)                 19.4   30.0
denote          Dhall.Core                 src/Dhall/Core.hs:(1248,1)-(1308,39)                 19.1   24.9
toHashMap       Data.HashMap.Strict.InsOrd src/Data/HashMap/Strict/InsOrd.hs:599:1-44           13.3   13.9
getInput        Text.Megaparsec            Text/Megaparsec.hs:404:1-40                           6.8    2.8
traverseWithKey Data.HashMap.Strict.InsOrd src/Data/HashMap/Strict/InsOrd.hs:(420,1)-(421,78)    4.9    2.9
normalize       Dhall.Core                 src/Dhall/Core.hs:1222:1-41                           4.1    7.1
new_            Data.HashMap.Array         Data/HashMap/Array.hs:177:1-28                        3.7    5.7
toList          Data.HashMap.Strict.InsOrd src/Data/HashMap/Strict/InsOrd.hs:(580,1)-(584,22)    2.2    0.4
```

The first commit, changing how `Let` is typechecked, is a factor of ~5 improvement; the second one, changing `toList`, is a factor of ~2.

ocharles commented on src/Dhall/TypeCheck.hs:
```diff
@@ -504,7 +523,7 @@ typeWithA tpa = loop
                     Const Kind -> return ()
                     _          -> Left (TypeError ctx e (InvalidAlternativeType k t))
-        mapM_ process (Data.HashMap.Strict.InsOrd.toList kts)
+        mapM_ process (Data.HashMap.Strict.toList (Data.HashMap.Strict.InsOrd.toHashMap kts))
```
Replacing this with `foldrWithKey` might be even better.

quasicomputational replied:
#419's benchmark. toList:

```
time                 974.7 ms   (922.6 ms .. 1.012 s)
                     1.000 R²   (0.999 R² .. NaN R²)
mean                 984.7 ms   (977.0 ms .. 989.4 ms)
std dev              7.569 ms   (2.795 ms .. 10.09 ms)
variance introduced by outliers: 19% (moderately inflated)
```

foldrWithKey:

```
time                 952.6 ms   (947.5 ms .. 957.4 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 953.1 ms   (951.8 ms .. 954.6 ms)
std dev              1.791 ms   (672.1 μs .. 2.307 ms)
variance introduced by outliers: 19% (moderately inflated)
```

A 3.3% gain. I guess since I've benchmarked it and found it to be a (small) win, I'll change it.

Gabriel439 commented on src/Dhall/TypeCheck.hs:
```diff
@@ -154,11 +155,29 @@ typeWithA tpa = loop
+          Const Type -> do
+            let ctx' = fmap (Dhall.Core.shift 1 (V x 0)) (Dhall.Context.insert x (Dhall.Core.normalize _A1) ctx)
+            _B0 <- loop ctx' b0
+            let a1 = Dhall.Core.shift 1 (V x 0) a0
```
Is there a reason why this does not normalize `a0` as in the catch-all branch?

Gabriel439 commented on src/Dhall/TypeCheck.hs:
```diff
@@ -154,11 +155,29 @@ typeWithA tpa = loop
+        -- and then don't need to again at the use sites; this can be
+        -- a very significant speed-up - see #412.
+        case Dhall.Core.normalize t of
+          Const Type -> do
```
Have you tried doing this fast-path for all expressions (i.e. not just terms)?

Gabriel439 replied:
Oh, never mind.  I see from your comment on the original issue that you don't want to do this for types because then it breaks support for type synonyms

Note that this implies that if we ever turn on true dependent types (i.e. functions from terms to types) then we would have to eliminate this fast path.  Not that I plan on introducing dependent type support any time soon, but it's worth adding a comment to this line:

https://github.com/dhall-lang/dhall-haskell/blob/6f626c96bcb79ea7d1b421487fad69392ad7e8c0/src/Dhall/TypeCheck.hs#L66

... explaining that if we turn on dependent types then we need to turn off this fast path to avoid a type-checking bug

Gabriel439 commented on src/Dhall/TypeCheck.hs:
```diff
@@ -154,11 +155,29 @@ typeWithA tpa = loop
+
+        t <- loop ctx _A1
+
+        -- The catch-all branch will actually do the correct thing for
```
Perhaps rephrase this comment to explain that the slow path (i.e. the catch-all branch) exactly matches the standard and has to be used for right-hand-sides that are types in order to correctly support type synonyms.  For right-hand-sides that are not types we use the fast path which doesn't require type-checking the right-hand-side of the `let` expression multiple times if the bound variable is referenced repeatedly

quasicomputational commented:
I added the dropped `normalize` and now the branches are obviously quite parallel, and I rewrote the comment and added another to, hopefully, be clearer.

quasicomputational commented:
I just pushed another commit that helps matters. Before, on `dhall-to-cabal.dhall`:

```
	   dhall-hash +RTS -p -RTS

	total time  =       15.45 secs   (15446 ticks @ 1000 us, 1 processor)
	total alloc = 8,419,758,564 bytes  (excludes profiling overheads)

COST CENTRE     MODULE                     SRC                                                 %time %alloc

getInput        Text.Megaparsec            Text/Megaparsec.hs:404:1-40                          28.0   21.4
traverseWithKey Data.HashMap.Strict.InsOrd src/Data/HashMap/Strict/InsOrd.hs:(420,1)-(421,78)   19.0    8.9
normalizeWith   Dhall.Core                 src/Dhall/Core.hs:(1325,1)-(1675,22)                 13.9   21.5
denote          Dhall.Core                 src/Dhall/Core.hs:(1247,1)-(1307,39)                 12.7   17.3
toHashMap       Data.HashMap.Strict.InsOrd src/Data/HashMap/Strict/InsOrd.hs:599:1-44            9.2   10.4
normalize       Dhall.Core                 src/Dhall/Core.hs:1221:1-41                           2.9    6.0
toList          Data.HashMap.Strict.InsOrd src/Data/HashMap/Strict/InsOrd.hs:(580,1)-(584,22)    2.6    1.4
shift           Dhall.Core                 src/Dhall/Core.hs:(647,1)-(800,29)                    2.6    3.9
new_            Data.HashMap.Array         Data/HashMap/Array.hs:177:1-28                        2.5    4.7
typeWithA       Dhall.TypeCheck            src/Dhall/TypeCheck.hs:(102,1)-(758,49)               2.2    0.8
```

```
benchmarking issue 412
time                 923.2 ms   (897.1 ms .. 942.8 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 929.3 ms   (924.9 ms .. 932.6 ms)
std dev              4.235 ms   (216.8 μs .. 5.160 ms)
variance introduced by outliers: 19% (moderately inflated)
```

After:

```
	   dhall-hash +RTS -p -RTS

	total time  =       12.68 secs   (12684 ticks @ 1000 us, 1 processor)
	total alloc = 7,429,890,836 bytes  (excludes profiling overheads)

COST CENTRE     MODULE                     SRC                                                 %time %alloc

getInput        Text.Megaparsec            Text/Megaparsec.hs:404:1-40                          32.1   24.3
normalizeWith   Dhall.Core                 src/Dhall/Core.hs:(1325,1)-(1675,22)                 15.6   22.9
denote          Dhall.Core                 src/Dhall/Core.hs:(1247,1)-(1307,39)                 15.4   19.6
traverseWithKey Data.HashMap.Strict.InsOrd src/Data/HashMap/Strict/InsOrd.hs:(420,1)-(421,78)   15.1    6.8
toHashMap       Data.HashMap.Strict.InsOrd src/Data/HashMap/Strict/InsOrd.hs:599:1-44            6.3    7.7
shift           Dhall.Core                 src/Dhall/Core.hs:(647,1)-(800,29)                    3.2    4.4
new_            Data.HashMap.Array         Data/HashMap/Array.hs:177:1-28                        2.8    4.5
normalize       Dhall.Core                 src/Dhall/Core.hs:1221:1-41                           2.0    4.4
toList          Data.HashMap.Strict.InsOrd src/Data/HashMap/Strict/InsOrd.hs:(580,1)-(584,22)    1.8    1.0
typeWithA       Dhall.TypeCheck            src/Dhall/TypeCheck.hs:(102,1)-(757,49)               1.5    0.7
```

```
benchmarking issue 412
time                 893.0 ms   (870.8 ms .. 919.5 ms)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 894.6 ms   (889.8 ms .. 898.3 ms)
std dev              4.839 ms   (3.211 ms .. 5.868 ms)
variance introduced by outliers: 19% (moderately inflated)
```

quasicomputational commented:
I'll continue pondering it a bit and see if any more easy optimisations will drop out. `dhall-to-cabal` is back to sort of acceptable performance with these patches (actually better than it's been in a while, despite the 300-constructor union that's been added recently); thanks for the tips and review!

Gabriel439 commented:
@quasicomputational: You're welcome!  Also, thanks for the performance improvements :)
